src/Fable.React.Helpers.fs

```diff
@@ -173,18 +173,56 @@ module Helpers =
 
     [<Emit("typeof $0 === 'object' && !$0[Symbol.iterator]")>]
     let private isNonEnumerableObject (x: obj): bool = jsNative
+
+    [<Emit("$0.length")>]
+    let private jsArrayLength(arr: ResizeArray<_>): int = jsNative
 #endif
 
     /// Same as F# equality but ignores functions in the first level of an object
-    /// Useful in combination with memoBuilderWith for most cases (ignore Elmish dispatch, etc)
+    /// Useful in combination with `FunctionComponent.Of` `memoizeWith` parameter for most cases
+    /// (ignore Elmish dispatch, etc)
     let equalsButFunctions (x: 'a) (y: 'a) =
 #if FABLE_COMPILER
         if obj.ReferenceEquals(x, y) then
             true
         elif isNonEnumerableObject x && not(isNull(box y)) then
-            (true, JS.Object.keys (x)) ||> Seq.fold (fun eq k ->
-                eq && (isFunction x?(k) || x?(k) = y?(k)))
-        else (box x) = (box y)
+            let keys = JS.Object.keys x
+            let length = jsArrayLength(keys)
+            let mutable i = 0
+            let mutable result = true
+            while i < length && result do
+                let key = keys.[i]
+                i <- i + 1
+                let xValue = x?(key)
+                result <- isFunction xValue || xValue = y?(key)
+            result
+        else
+            (box x) = (box y)
+#else
+        // Server rendering, won't be actually used
+        // Avoid `x = y` because it will force 'a to implement structural equality
+        false
+#endif
+
+    /// Same as the default `React.memo` equality but ignores functions.
+    /// Useful in combination with `FunctionComponent.Of` `memoizeWith` parameter when F# equality isn't required
+    let memoEqualsButFunctions (x: 'a) (y: 'a) =
```

Comment:
If you don't mind, we could change the name to `refEqualsButFunctions` to make the semantics clearer. I'm afraid it may not be immediately obvious to users this is closer to how React.memo default equality works (and they may not know what that means).

Reply:
That was my first name but I feared that reference equality would be misleading.

The name doesn't exists in JS so JS/React developers wouldn't guess it and as it uses `===` it's actually not doing a reference equality check, strings in particular are compared by value (something we very much want)

Memo isn't perfect because it assumes JS/React knowledge either but it sounded less of a subtle lie.

Anyway feel free to change, I don't think there is any perfect answer there it's on the fence between JS and .NET communities :)

Reply:
Hmm, you're right. Actually you made me think because in Fable we translate `obj.ReferenceEquals` as `===` but this gives different results in .NET and Fable for value types and strings:

```fsharp
let x = 5
let y = 5

// false in .NET, true in Fable
obj.ReferenceEquals(x, y) |> printfn "%b"
```

Should we use another name, `primitiveEquals` or similar?

Reply:
I think it's ok that obj.ReferenceEquals do is as AFAIK true reference equality isn't avaiable in JS so it's the nearest thing.

`primitiveEquals` seem good as it doesn't have the mental association with .NET reference equality

Reply:
It seems the closes function in F# to JS `===` behavior is [LanguagePrimitives.PhysicalEquality](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/languageprimitives.physicalequality%5B't%5D-function-%5Bfsharp%5D). Should we use `physicalEquality`?

I doubt it, but we also should be careful in case obj.ReferenceEquals implementation in Fable changes in the future. Maybe we should just write the function in JS at the end :/ And also probably you're right `memoEquals` may be the best name ;)